kernel/src/process_load.rs

```diff
@@ -0,0 +1,741 @@
+            },
+
+            ProcessLoadError::InternalError => write!(f, "Error in kernel. Likely a bug."),
+
+            ProcessLoadError::CredentialsNoAccept => write!(f, "No credentials accepted."),
+
+            ProcessLoadError::CredentialsReject(index) => {
+                write!(f, "Credentials index {} rejected.", index)
+            }
+        }
+    }
+}
+
+/// Load processes (stored as TBF objects in flash) into runnable
+/// process structures stored in the `procs` array. If a `checker` is
+/// passed, this method scans the footers in the TBF for cryptographic
+/// credentials for binary integrity, passing them to `checker` to decide
+/// whether the process has sufficient credentials to run.
+#[inline(always)]
+pub fn load_and_check_processes<C: Chip>(
+    kernel: &'static Kernel,
+    chip: &'static C,
+    app_flash: &'static [u8],
+    app_memory: &'static mut [u8],
+    mut procs: &'static mut [Option<&'static dyn Process>],
+    fault_policy: &'static dyn ProcessFaultPolicy,
+    capability_management: &dyn ProcessManagementCapability,
+) -> Result<(), ProcessLoadError> {
+    load_processes(
+        kernel,
+        chip,
+        app_flash,
+        app_memory,
+        &mut procs,
+        fault_policy,
+        capability_management,
+    )?;
+    let _res = check_processes(procs, kernel);
+    Ok(())
+}
+
+/// Helper function to load processes from flash into an array of active
+/// processes. This is the default template for loading processes, but a board
+/// is able to create its own `load_processes()` function and use that instead.
+///
+/// Processes are found in flash starting from the given address and iterating
+/// through Tock Binary Format (TBF) headers. Processes are given memory out of
+/// the `app_memory` buffer until either the memory is exhausted or the
+/// allocated number of processes are created. This buffer is a non-static slice,
+/// ensuring that this code cannot hold onto the slice past the end of this function
+/// (instead, processes store a pointer and length), which necessary for later
+/// creation of `ProcessBuffer`s in this memory region to be sound.
+/// A reference to each process is stored in the provided `procs` array.
+/// How process faults are handled by the
+/// kernel must be provided and is assigned to every created process.
+///
+/// This function is made `pub` so that board files can use it, but loading
+/// processes from slices of flash an memory is fundamentally unsafe. Therefore,
+/// we require the `ProcessManagementCapability` to call this function.
+///
+/// Returns `Ok(())` if process discovery went as expected. Returns a
+/// `ProcessLoadError` if something goes wrong during TBF parsing or process
+/// creation.
+#[inline(always)]
+fn load_processes<C: Chip>(
+    kernel: &'static Kernel,
+    chip: &'static C,
+    app_flash: &'static [u8],
+    app_memory: &'static mut [u8],
+    procs: &mut &'static mut [Option<&'static dyn Process>],
```

Comment:
It's because the method needs to hold onto the reference after it returns; in the previous formulation, `process_load` held onto `procs` only for its duration (hence the anonymous lifetime). Since the method now starts an asynchronous state machine, you need to pass the underlying reference.